lib/BackgroundJob/SendScheduledMessages.php

```diff
@@ -0,0 +1,214 @@
+					$message->getActorId()
+				);
+				continue;
+			}
+
+			if ($room->isFederatedConversation()) {
+				$this->logger->warning('Cannot send scheduled messages to federated conversations');
+				$this->scheduledMessageService->deleteMessage(
+					$room,
+					(string)$message->getId(),
+					$participant->getAttendee()->getActorType(),
+					$participant->getAttendee()->getActorId()
+				);
+				$this->participantService->setHasScheduledMessages($participant, Participant::ERROR_SCHEDULED_MESSAGE);
+				continue;
+			}
+
+			if ($room->getReadOnly() === Room::READ_ONLY) {
+				$this->logger->warning('Cannot send scheduled messages to read only room');
+				$this->scheduledMessageService->markAsFailed($message);
+				$this->participantService->setHasScheduledMessages($participant, Participant::ERROR_SCHEDULED_MESSAGE);
+				continue;
+			}
+
+			if ($room->getLobbyState() !== Webinary::LOBBY_NONE && ($participant->getPermissions() & Attendee::PERMISSIONS_LOBBY_IGNORE) === 0) {
+				$this->logger->warning('User has no chat permissions for room ' . $message->getRoomId() . ', could not send scheduled message');
+				$this->scheduledMessageService->markAsFailed($message);
+				$this->participantService->setHasScheduledMessages($participant, Participant::ERROR_SCHEDULED_MESSAGE);
+				continue;
+			}
+
+			if (($participant->getPermissions() & Attendee::PERMISSIONS_CHAT) === 0) {
+				$this->logger->warning('User has no chat permissions for room ' . $message->getRoomId() . ', could not send scheduled message');
+				$this->scheduledMessageService->markAsFailed($message);
+				$this->participantService->setHasScheduledMessages($participant, Participant::ERROR_SCHEDULED_MESSAGE);
+				continue;
+			}
+
+			$parent = $parentMessage = null;
+			if ($message->getParentId() !== 0 && $message->getParentId() !== null) {
+				try {
+					$parent = $this->chatManager->getParentComment($room, (string)$message->getParentId());
+					$parentMessage = $this->messageParser->createMessage($room, $participant, $parent, $this->l10n);
+					$this->messageParser->parseMessage($parentMessage);
+					if (!$parentMessage->isReplyable()) {
+						$this->logger->warning('Parent message for scheduled message not replyable for message ' . $message->getId());
+						$this->scheduledMessageService->markAsFailed($message);
+						$this->participantService->setHasScheduledMessages($participant, Participant::ERROR_SCHEDULED_MESSAGE);
+						continue;
+					}
+				} catch (NotFoundException $e) {
+					$this->logger->warning('Parent for scheduled message not found, sending without parent', ['exception' => $e]);
+					$message->setParentId(null);
+				}
+			} elseif ($message->getThreadId() !== 0 && $message->getThreadId() !== -1) {
+				if (!$this->threadService->validateThread($room->getId(), $message->getThreadId())) {
+					$message->setThreadId(0);
+					$this->logger->warning('Could not validate thread for scheduled message, set thread to 0 for ' . $message->getId());
```

Comment:
that id is the scheduled one which is not available afterwards anymore?
Since it could lack context, should we instead fail it?
Same for replying?